pkg/util/template/funcerror_test.go

```diff
@@ -0,0 +1,72 @@
+package template
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/authgear/authgear-server/pkg/api/apierrors"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestFuncsError(t *testing.T) {
+	makeApiErr := func(jsonStr string) *apierrors.APIError {
+		var apiErr *apierrors.APIError
+		err := json.Unmarshal([]byte(jsonStr), &apiErr)
+		if err != nil {
+			panic(err)
+		}
+		return apiErr
+	}
+	Convey("Resolve validation required error", t, func() {
+		Convey("newPasswordField error", func() {
+			apiErr := makeApiErr(`
+{
+	"code": 400,
+	"info": {
+		"causes": [
+			{
+				"details": {
+					"actual": [
+						"gorilla.csrf.Token",
+						"x_step"
+					],
+					"expected": [
+						"x_confirm_password",
+						"x_password"
+					],
+					"missing": [
+						"x_confirm_password",
+						"x_password"
+					]
+				},
+				"kind": "required",
+				"location": ""
+			}
+		]
+	},
+	"message": "invalid value",
+	"name": "Invalid",
+	"reason": "ValidationFailed"
+}
+`)
```

Comment:
Given

```
{
    "code": 400,
    "info": {
        "causes": [
            {
                "details": {
                    "actual": [
                        "gorilla.csrf.Token",
                        "x_step"
                    ],
                    "expected": [
                        "x_confirm_password",
                        "x_password"
                    ],
                    "missing": [
                        "x_confirm_password",
                        "x_password"
                    ]
                },
                "kind": "required",
                "location": ""
            },
            {
              "kind": "minLength",
              "location": "/x_password"
            },
            {
              "kind": "minLength",
              "location": "/x_confirm_password"
            }
        ]
    },
    "message": "invalid value",
    "name": "Invalid",
    "reason": "ValidationFailed"
}
```

```
resolveError $.RawError (dict
  newPasswordField (dict
    "by_location" (list "x_password")
  )
)
```

it should return

```
{
    "code": 400,
    "info": {
        "causes": [
            {
                "details": {
                    "actual": [
                        "gorilla.csrf.Token",
                        "x_step"
                    ],
                    "expected": [
                        "x_password"
                    ],
                    "missing": [
                        "x_password"
                    ]
                },
                "kind": "required",
                "location": ""
            },
            {
              "kind": "minLength",
              "location": "/x_password"
            }
        ]
    },
    "message": "invalid value",
    "name": "Invalid",
    "reason": "ValidationFailed"
}

```

---

Problems in your proposed API

1. Things should be optional. If I do not care about reason, I do not need to specify `by_reason`
2. You still did not "slice" the errors. See my above expected output.

Reply:
okok! Lemme update it